README
NAME
    Time::Local::Extended - Increase the range of localtime and timelocal

SYNOPSIS
      use Time::Local::Extended qw(:ALL);

      my @localtime   = localtime(2**31);
      my $seconds     = timelocal(0,0,0,1,10,170);
      my $gmt_seconds = timegm(0,0,0,1,10,170);
      my $gmt_time    = gmtime(2**31);
      my $ux_time     = UNIX_TIMESTAMP('2097-07-04 12:34:56');
      my $date        = FROM_UNIXTIME(2**31);

      my $sql = qq(
             SELECT start_time
             FROM   project
             WHERE  id = 1
      );
      my $date_time = $dbh->selectrow_array($sql); # '2097-07-04 12:34:56'
      my $ux_time   = UNIX_TIMESTAMP($date_time);  # 4023794096

      my $date_time = FROM_UNIXTIME(2**31);
      my $sql  = qq(
             UPDATE project
             SET    start_time = ?
             WHERE  id = 1
      );
      $dbh->do($sql, undef, $date_time);

DESCRIPTION
    This module extends the date range of localtime(), gmtime(), timegm()
    and timelocal() to go safely beyond 2038 and before 1970 on any
    operating system.

    It also provides a handful of useful time conversion functions.

PUBLIC FUNCTIONS
    *   timelocal

        Invoked in the same way as Time::Local::timelocal().

    *   localtime

        Invoked in the same way as CORE::localtime().

    *   timegm

        Invoked in the same way as Time::Local::timegm().

    *   gmtime

        Invoked in the same way as CORE::gmtime().

    *   UNIX_TIMESTAMP

        Invoked similarly to the MySQL UNIX_TIMESTAMP() function()

    *   FROM_UNIXTIME

        Invoked similarly to the MySQL FROM_UNIXTIME() function()

LIMITATIONS
    Because of the way timegm() and timelocal() try to Do What You Mean with
    the year, it is impossible to feed it the years 0 through 99 (it thinks
    you mean 2000 - 2099).

    While the code can in theory go out to 2**63, the practical portable
    limit of this code is from 2**52 to -2**52 (the limit of double floating
    point precision) after which precision starts to drop off. This gives
    you a range of about +/- 142 million years.

BUGS
    Doesn't correctly handle British Summer Time 1968-1971

    Please e-mail bug reports or suggestions to bobo@cpan.org. Thanks!

CREDITS
    Thanks to Michael Schwern for extending this module's capabilities
    beyond 2098.

    Thanks to Peter Kioko for helping to refine the idea, and Adam Foxson,
    the Human CPAN Reference Manual, for quality assurance.

AUTHOR
    Bob O'Neill, <bobo@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2003-2008 Bob O'Neill and Michael Schwern. All rights
    reserved.

    See COPYING for license

SEE ALSO
    *   perl.

    *   Time::Local.



Changes
Revision history for Perl extension Time::Local::Extended.

0.50  Oct 18, 2008

      Schwernifications:

      New Features
      * Now accurate beyond 2090 and before 1900.

      Bug Fixes
      * localtime() and timelocal() are now more accurate about DST changes.

      Misc
      * Now using the y2038 project

0.48  Mon Jun 28 22:30:00 2004 EDT

     - Fixed skip() in tests to agree with number of planned tests.

0.47  Sat Jun 26 18:00:00 2004 EDT

     - Fix for daylight savings time, thanks to Vytautas Mickevicius.
     - Updated copyright.

0.44  Sun Apr 04 22:30:00 2004 EDT

     - Added exemption for non-Eastern time zones in t/export_tests.t

0.43  Sun Apr 04 20:00:00 2004 EST

     - Added support for specifying 4-digit years to timelocal.
       Thanks to Vytautas Mickevicius for the bug report.
     - Added support for the 2-digit year "rolling window" to syncrhonize
       functionality with Time::Local::timelocal.  Algorithm adapted from
       Time::Local.
     - Added localtime and gmtime to functions exported by default since
       it makes sense for localtime to be in sync with timelocal.

0.41  Wed Nov 05 22:00:00 2003 EST

     - Added support for dates up to 2098.
     - Renamed distribution as Time::Local::Extended.
     - Changed @EXPORT to be the same as @Time::Local::EXPORT.
       Thanks to Michael Schwern for this advice.
     - Documentation updated.
     - Tests updated.
     - Corrected weekday of first release in Changes file.

0.33  Wed Nov 05 00:00:00 2003 EST
     - initial release (as Time::Local::2074)



time64.c
/* 

Copyright (c) 2007-2008  Michael G Schwern

This software originally derived from Paul Sheer's pivotal_gmtime_r.c.

The MIT License:

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

*/

/*

Programmers who have available to them 64-bit time values as a 'long
long' type can use localtime64_r() and gmtime64_r() which correctly
converts the time even on 32-bit systems. Whether you have 64-bit time 
values will depend on the operating system.

localtime64_r() is a 64-bit equivalent of localtime_r().

gmtime64_r() is a 64-bit equivalent of gmtime_r().

*/

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <errno.h>
#include "time64.h"

static const int days_in_month[2][12] = {
    {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31},
    {31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31},
};

static const int julian_days_by_month[2][12] = {
    {0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334},
    {0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335},
};

static const int length_of_year[2] = { 365, 366 };

/* Number of days in a 400 year Gregorian cycle */
static const Year years_in_gregorian_cycle = 400;
static const int days_in_gregorian_cycle  = (365 * 400) + 100 - 4 + 1;

/* 28 year calendar cycle between 2010 and 2037 */
#define SOLAR_CYCLE_LENGTH 28
static const int safe_years[SOLAR_CYCLE_LENGTH] = {
    2016, 2017, 2018, 2019,
    2020, 2021, 2022, 2023,
    2024, 2025, 2026, 2027,
    2028, 2029, 2030, 2031,
    2032, 2033, 2034, 2035,
    2036, 2037, 2010, 2011,
    2012, 2013, 2014, 2015
};

static const int dow_year_start[SOLAR_CYCLE_LENGTH] = {
    5, 0, 1, 2,     /* 0       2016 - 2019 */
    3, 5, 6, 0,     /* 4  */
    1, 3, 4, 5,     /* 8  */
    6, 1, 2, 3,     /* 12 */
    4, 6, 0, 1,     /* 16 */
    2, 4, 5, 6,     /* 20      2036, 2037, 2010, 2011 */
    0, 2, 3, 4      /* 24      2012, 2013, 2014, 2015 */
};

/* Let's assume people are going to be looking for dates in the future.
   Let's provide some cheats so you can skip ahead.
   This has a 4x speed boost when near 2008.
*/
/* Number of days since epoch on Jan 1st, 2008 GMT */
#define CHEAT_DAYS  (1199145600 / 24 / 60 / 60)
#define CHEAT_YEARS 108

#define IS_LEAP(n)	((!(((n) + 1900) % 400) || (!(((n) + 1900) % 4) && (((n) + 1900) % 100))) != 0)
#define WRAP(a,b,m)	((a) = ((a) <  0  ) ? ((b)--, (a) + (m)) : (a))

#ifdef USE_SYSTEM_LOCALTIME
#    define SHOULD_USE_SYSTEM_LOCALTIME(a)  (       \
    (a) <= SYSTEM_LOCALTIME_MAX &&              \
    (a) >= SYSTEM_LOCALTIME_MIN                 \
)
#else
#    define SHOULD_USE_SYSTEM_LOCALTIME(a)      (0)
#endif

#ifdef USE_SYSTEM_GMTIME
#    define SHOULD_USE_SYSTEM_GMTIME(a)     (       \
    (a) <= SYSTEM_GMTIME_MAX    &&              \
    (a) >= SYSTEM_GMTIME_MIN                    \
)
#else
#    define SHOULD_USE_SYSTEM_GMTIME(a)         (0)
#endif

/* Multi varadic macros are a C99 thing, alas */
#ifdef TIME_64_DEBUG
#    define TRACE(format) (fprintf(stderr, format))
#    define TRACE1(format, var1)    (fprintf(stderr, format, var1))
#    define TRACE2(format, var1, var2)    (fprintf(stderr, format, var1, var2))
#    define TRACE3(format, var1, var2, var3)    (fprintf(stderr, format, var1, var2, var3))
#else
#    define TRACE(format) ((void)0)
#    define TRACE1(format, var1) ((void)0)
#    define TRACE2(format, var1, var2) ((void)0)
#    define TRACE3(format, var1, var2, var3) ((void)0)
#endif

static int is_exception_century(Year year)
{
    int is_exception = ((year % 100 == 0) && !(year % 400 == 0));
    TRACE1("# is_exception_century: %s\n", is_exception ? "yes" : "no");

    return(is_exception);
}


Time64_T timegm64(struct TM *date) {
    int      days    = 0;
    Time64_T seconds = 0;
    Year     year;

    if( date->tm_year > 70 ) {
        year = 70;
        while( year < date->tm_year ) {
            days += length_of_year[IS_LEAP(year)];
            year++;
        }
    }
    else if ( date->tm_year < 70 ) {
        year = 69;
        do {
            days -= length_of_year[IS_LEAP(year)];
            year--;
        } while( year >= date->tm_year );
    }

    days += julian_days_by_month[IS_LEAP(date->tm_year)][date->tm_mon];
    days += date->tm_mday - 1;

    /* Avoid overflowing the days integer */
    seconds = days;
    seconds = seconds * 60 * 60 * 24;

    seconds += date->tm_hour * 60 * 60;
    seconds += date->tm_min * 60;
    seconds += date->tm_sec;

    return(seconds);
}


static int check_tm(struct TM *tm)
{
    /* Don't forget leap seconds */
    assert(tm->tm_sec >= 0);
    assert(tm->tm_sec <= 61);

    assert(tm->tm_min >= 0);
    assert(tm->tm_min <= 59);

    assert(tm->tm_hour >= 0);
    assert(tm->tm_hour <= 23);

    assert(tm->tm_mday >= 1);
    assert(tm->tm_mday <= days_in_month[IS_LEAP(tm->tm_year)][tm->tm_mon]);

    assert(tm->tm_mon  >= 0);
    assert(tm->tm_mon  <= 11);

    assert(tm->tm_wday >= 0);
    assert(tm->tm_wday <= 6);
    
    assert(tm->tm_yday >= 0);
    assert(tm->tm_yday <= length_of_year[IS_LEAP(tm->tm_year)]);

#ifdef HAS_TM_TM_GMTOFF
    assert(tm->tm_gmtoff >= -24 * 60 * 60);
    assert(tm->tm_gmtoff <=  24 * 60 * 60);
#endif

    return 1;
}


/* The exceptional centuries without leap years cause the cycle to
   shift by 16
*/
static Year cycle_offset(Year year)
{
    const Year start_year = 2000;
    Year year_diff  = year - start_year;
    Year exceptions;

    if( year > start_year )
        year_diff--;

    exceptions  = year_diff / 100;
    exceptions -= year_diff / 400;

    TRACE3("# year: %lld, exceptions: %lld, year_diff: %lld\n",
          year, exceptions, year_diff);

    return exceptions * 16;
}

/* For a given year after 2038, pick the latest possible matching
   year in the 28 year calendar cycle.

   A matching year...
   1) Starts on the same day of the week.
   2) Has the same leap year status.

   This is so the calendars match up.

   Also the previous year must match.  When doing Jan 1st you might
   wind up on Dec 31st the previous year when doing a -UTC time zone.

   Finally, the next year must have the same start day of week.  This
   is for Dec 31st with a +UTC time zone.
   It doesn't need the same leap year status since we only care about
   January 1st.
*/
int safe_year(Year year)
{
    int safe_year;
    Year year_cycle = year + cycle_offset(year);

    /* Change non-leap xx00 years to an equivalent */
    if( is_exception_century(year) )
        year_cycle += 11;

    /* Also xx01 years, since the previous year will be wrong */
    if( is_exception_century(year - 1) )
        year_cycle += 17;

    year_cycle %= SOLAR_CYCLE_LENGTH;
    if( year_cycle < 0 ) 
        year_cycle = SOLAR_CYCLE_LENGTH + year_cycle;

    assert( year_cycle >= 0 );
    assert( year_cycle < SOLAR_CYCLE_LENGTH );
    safe_year = safe_years[year_cycle];

    assert(safe_year <= 2037 && safe_year >= 2010);
    
    TRACE3("# year: %lld, year_cycle: %lld, safe_year: %d\n",
          year, year_cycle, safe_year);

    return safe_year;
}


void copy_tm_to_TM(const struct tm *src, struct TM *dest) {
    if( src == NULL ) {
        memset(dest, 0, sizeof(*dest));
    }
    else {
#       ifdef USE_TM64
            dest->tm_sec        = src->tm_sec;
            dest->tm_min        = src->tm_min;
            dest->tm_hour       = src->tm_hour;
            dest->tm_mday       = src->tm_mday;
            dest->tm_mon        = src->tm_mon;
            dest->tm_year       = (Year)src->tm_year;
            dest->tm_wday       = src->tm_wday;
            dest->tm_yday       = src->tm_yday;
            dest->tm_isdst      = src->tm_isdst;

#           ifdef HAS_TM_TM_GMTOFF
                dest->tm_gmtoff  = src->tm_gmtoff;
#           endif

#           ifdef HAS_TM_TM_ZONE
                dest->tm_zone  = src->tm_zone;
#           endif

#       else
            /* They're the same type */
            memcpy(dest, src, sizeof(*dest));
#       endif
    }
}


void copy_TM_to_tm(const struct TM *src, struct tm *dest) {
    if( src == NULL ) {
        memset(dest, 0, sizeof(*dest));
    }
    else {
#       ifdef USE_TM64
            dest->tm_sec        = src->tm_sec;
            dest->tm_min        = src->tm_min;
            dest->tm_hour       = src->tm_hour;
            dest->tm_mday       = src->tm_mday;
            dest->tm_mon        = src->tm_mon;
            dest->tm_year       = (int)src->tm_year;
            dest->tm_wday       = src->tm_wday;
            dest->tm_yday       = src->tm_yday;
            dest->tm_isdst      = src->tm_isdst;

#           ifdef HAS_TM_TM_GMTOFF
                dest->tm_gmtoff  = src->tm_gmtoff;
#           endif

#           ifdef HAS_TM_TM_ZONE
                dest->tm_zone  = src->tm_zone;
#           endif

#       else
            /* They're the same type */
            memcpy(dest, src, sizeof(*dest));
#       endif
    }
}


/* Simulate localtime_r() to the best of our ability */
struct tm * fake_localtime_r(const time_t *clock, struct tm *result) {
    const struct tm *static_result = localtime(clock);

    assert(result != NULL);

    if( static_result == NULL ) {
        memset(result, 0, sizeof(*result));
        return NULL;
    }
    else {
        memcpy(result, static_result, sizeof(*result));
        return result;
    }
}


/* Simulate gmtime_r() to the best of our ability */
struct tm * fake_gmtime_r(const time_t *clock, struct tm *result) {
    const struct tm *static_result = gmtime(clock);

    assert(result != NULL);

    if( static_result == NULL ) {
        memset(result, 0, sizeof(*result));
        return NULL;
    }
    else {
        memcpy(result, static_result, sizeof(*result));
        return result;
    }
}


struct TM *gmtime64_r (const Time64_T *in_time, struct TM *p)
{
    int v_tm_sec, v_tm_min, v_tm_hour, v_tm_mon, v_tm_wday;
    Time64_T v_tm_tday;
    int leap;
    Time64_T m;
    Time64_T time = *in_time;
    Year year = 70;
    int cycles = 0;

    assert(p != NULL);

    /* Use the system gmtime() if time_t is small enough */
    if( SHOULD_USE_SYSTEM_GMTIME(*in_time) ) {
        time_t safe_time = *in_time;
        struct tm safe_date;
        GMTIME_R(&safe_time, &safe_date);

        copy_tm_to_TM(&safe_date, p);
        assert(check_tm(p));

        return p;
    }

#ifdef HAS_TM_TM_GMTOFF
    p->tm_gmtoff = 0;
#endif
    p->tm_isdst  = 0;

#ifdef HAS_TM_TM_ZONE
    p->tm_zone   = "UTC";
#endif

    v_tm_sec =  (int)(time % 60);
    time /= 60;
    v_tm_min =  (int)(time % 60);
    time /= 60;
    v_tm_hour = (int)(time % 24);
    time /= 24;
    v_tm_tday = time;

    WRAP (v_tm_sec, v_tm_min, 60);
    WRAP (v_tm_min, v_tm_hour, 60);
    WRAP (v_tm_hour, v_tm_tday, 24);

    v_tm_wday = (int)((v_tm_tday + 4) % 7);
    if (v_tm_wday < 0)
        v_tm_wday += 7;
    m = v_tm_tday;

    if (m >= CHEAT_DAYS) {
        year = CHEAT_YEARS;
        m -= CHEAT_DAYS;
    }

    if (m >= 0) {
        /* Gregorian cycles, this is huge optimization for distant times */
        cycles = (int)(m / (Time64_T) days_in_gregorian_cycle);
        if( cycles ) {
            m -= (cycles * (Time64_T) days_in_gregorian_cycle);
            year += (cycles * years_in_gregorian_cycle);
        }

        /* Years */
        leap = IS_LEAP (year);
        while (m >= (Time64_T) length_of_year[leap]) {
            m -= (Time64_T) length_of_year[leap];
            year++;
            leap = IS_LEAP (year);
        }

        /* Months */
        v_tm_mon = 0;
        while (m >= (Time64_T) days_in_month[leap][v_tm_mon]) {
            m -= (Time64_T) days_in_month[leap][v_tm_mon];
            v_tm_mon++;
        }
    } else {
        year--;

        /* Gregorian cycles */
        cycles = (int)((m / (Time64_T) days_in_gregorian_cycle) + 1);
        if( cycles ) {
            m -= (cycles * (Time64_T) days_in_gregorian_cycle);
            year += (cycles * years_in_gregorian_cycle);
        }

        /* Years */
        leap = IS_LEAP (year);
        while (m < (Time64_T) -length_of_year[leap]) {
            m += (Time64_T) length_of_year[leap];
            year--;
            leap = IS_LEAP (year);
        }

        /* Months */
        v_tm_mon = 11;
        while (m < (Time64_T) -days_in_month[leap][v_tm_mon]) {
            m += (Time64_T) days_in_month[leap][v_tm_mon];
            v_tm_mon--;
        }
        m += (Time64_T) days_in_month[leap][v_tm_mon];
    }

    p->tm_year = year;
    if( p->tm_year != year ) {
#ifdef EOVERFLOW
        errno = EOVERFLOW;
#endif
        return NULL;
    }

    /* At this point m is less than a year so casting to an int is safe */
    p->tm_mday = (int) m + 1;
    p->tm_yday = julian_days_by_month[leap][v_tm_mon] + (int)m;
    p->tm_sec  = v_tm_sec;
    p->tm_min  = v_tm_min;
    p->tm_hour = v_tm_hour;
    p->tm_mon  = v_tm_mon;
    p->tm_wday = v_tm_wday;
    
    assert(check_tm(p));

    return p;
}


struct TM *localtime64_r (const Time64_T *time, struct TM *local_tm)
{
    time_t safe_time;
    struct tm safe_date;
    struct TM gm_tm;
    Year orig_year;
    int month_diff;

    assert(local_tm != NULL);

    /* Use the system localtime() if time_t is small enough */
    if( SHOULD_USE_SYSTEM_LOCALTIME(*time) ) {
        safe_time = *time;

        TRACE1("Using system localtime for %lld\n", *time);

        LOCALTIME_R(&safe_time, &safe_date);

        copy_tm_to_TM(&safe_date, local_tm);
        assert(check_tm(local_tm));

        return local_tm;
    }

    if( gmtime64_r(time, &gm_tm) == NULL ) {
        TRACE1("gmtime64_r returned null for %lld\n", *time);
        return NULL;
    }

    orig_year = gm_tm.tm_year;

    if (gm_tm.tm_year > (2037 - 1900) ||
        gm_tm.tm_year < (1970 - 1900)
       )
    {
        TRACE1("Mapping tm_year %lld to safe_year\n", (Year)gm_tm.tm_year);
        gm_tm.tm_year = safe_year((Year)(gm_tm.tm_year + 1900)) - 1900;
    }

    safe_time = timegm64(&gm_tm);
    if( LOCALTIME_R(&safe_time, &safe_date) == NULL ) {
        TRACE1("localtime_r(%d) returned NULL\n", (int)safe_time);
        return NULL;
    }

    copy_tm_to_TM(&safe_date, local_tm);

    local_tm->tm_year = orig_year;
    if( local_tm->tm_year != orig_year ) {
        TRACE2("tm_year overflow: tm_year %lld, orig_year %lld\n",
              (Year)local_tm->tm_year, (Year)orig_year);

#ifdef EOVERFLOW
        errno = EOVERFLOW;
#endif
        return NULL;
    }


    month_diff = local_tm->tm_mon - gm_tm.tm_mon;

    /*  When localtime is Dec 31st previous year and
        gmtime is Jan 1st next year.
    */
    if( month_diff == 11 ) {
        local_tm->tm_year--;
    }

    /*  When localtime is Jan 1st, next year and
        gmtime is Dec 31st, previous year.
    */
    if( month_diff == -11 ) {
        local_tm->tm_year++;
    }

    /* GMT is Jan 1st, xx01 year, but localtime is still Dec 31st 
       in a non-leap xx00.  There is one point in the cycle
       we can't account for which the safe xx00 year is a leap
       year.  So we need to correct for Dec 31st comming out as
       the 366th day of the year.
    */
    if( !IS_LEAP(local_tm->tm_year) && local_tm->tm_yday == 365 )
        local_tm->tm_yday--;

    assert(check_tm(local_tm));
    
    return local_tm;
}


time64.h
#include <time.h>
#include "time64_config.h"

#ifndef TIME64_H
#    define TIME64_H


/* Set our custom types */
typedef INT_64_T        Int64;
typedef Int64           Time64_T;
typedef Int64           Year;


/* A copy of the tm struct but with a 64 bit year */
struct TM64 {
        int     tm_sec;
        int     tm_min;
        int     tm_hour;
        int     tm_mday;
        int     tm_mon;
        Year    tm_year;
        int     tm_wday;
        int     tm_yday;
        int     tm_isdst;

#ifdef HAS_TM_TM_GMTOFF
        long    tm_gmtoff;
#endif

#ifdef HAS_TM_TM_ZONE
        char    *tm_zone;
#endif
};


/* Decide which tm struct to use */
#ifdef USE_TM64
#define TM      TM64
#else
#define TM      tm
#endif   


/* Declare public functions */
struct TM *gmtime64_r    (const Time64_T *, struct TM *);
struct TM *localtime64_r (const Time64_T *, struct TM *);
Time64_T   timegm64      (struct TM *);

/* expose safe_year for our timelocal() hack */
int safe_year(Year);

/* Not everyone has gm/localtime_r(), provide a replacement */
#ifdef HAS_LOCALTIME_R
#    define LOCALTIME_R(clock, result) localtime_r(clock, result)
#else
#    define LOCALTIME_R(clock, result) fake_localtime_r(clock, result)
#endif
#ifdef HAS_GMTIME_R
#    define GMTIME_R(clock, result)    gmtime_r(clock, result)
#else
#    define GMTIME_R(clock, result)    fake_gmtime_r(clock, result)
#endif

#endif


Makefile.PL
use ExtUtils::MakeMaker;
use Config;

# To be portable.
my $objext = $Config{obj_ext};

die "\$Config{quadtype} must be defined for this module to work."
  unless $Config{quadtype};

WriteMakefile
(
    'NAME'         => 'Time::Local::Extended',
    'VERSION_FROM' => 'lib/Time/Local/Extended.pm',
    'PREREQ_PM'    => {
        'Time::Local' => 0,
    },
    OBJECT => "time64$objext Extended$objext",

    PL_FILES => {
        "munge_config" => "time64_config.h"
    },

    (
        $] >= 5.005 ?
        (
            ABSTRACT_FROM => 'lib/Time/Local/Extended.pm',
            AUTHOR        => 'Bob O\'Neill <bobo@cpan.org>',
        ) : ()
    ),

    clean => {
        FILES   => "time64_config.h"
    },
);

{
    package MY;

    sub MY::postamble {
        return <<"MAKE";
time64$objext : time64_config.h

MAKE
    }
}


lib/Time/Local/Extended.pm
package Time::Local::Extended;

use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);
use Exporter;
use DynaLoader;
use Carp qw(cluck);
use Time::Local qw();

@ISA         = qw(Exporter DynaLoader);
@EXPORT      = qw(timelocal localtime timegm gmtime);
@EXPORT_OK   = qw(UNIX_TIMESTAMP FROM_UNIXTIME);
%EXPORT_TAGS = (ALL => [@EXPORT, @EXPORT_OK]);
$VERSION     = '0.51';
local $^W    = 1;

bootstrap Time::Local::Extended $VERSION;

# Follow the Time::Local::timelocal conventions:
#
# 1) Treat years greater than 999 as the actual 4-digit year
#    (not an offset from 1900)
# 2) Treat years in the range 0..99 as years in the "current century"
sub munge_year {
    my $year  = shift;

    return $year - 1900 if $year > 999 or $year < 0;

    if ($year >= 0 and $year <= 99)
    {
        my $current_year    = (CORE::localtime())[5] + 1900;
        my $current_century = int ($current_year / 100) * 100;
 
        my $break_point = $current_year + 50;
           $current_century += 100 if ($break_point % 100) < 50;

        my $adjusted_year  = $current_century + $year;
           $adjusted_year -= 100 if ($year + $current_century) > $break_point;

        return $adjusted_year - 1900;
    }

    return $year;
}

sub timelocal
{
    my @time_data = @_;

    $time_data[5] = munge_year($time_data[5]);

    # Need to adjust if year is 2038 or greater, even in January, because
    # Time::Local::timelocal() breaks at Jan 1 2038 rather than Jan 18, 2038.
    # Also if the year is at or before 1970 because negative times often don't
    # work.
    my $adjusting = ($time_data[5] >= 138 or $time_data[5] <= 70) ? 1 : 0;
    my $num_years = 0;

    my $orig_year;
    my $safe_year;
    my @adjusted_time = @time_data;
    if ($adjusting)
    {
        $orig_year = $time_data[5] + 1900;
        $safe_year = safe_year($orig_year);
        $num_years = $orig_year - $safe_year;
        $adjusted_time[5] = $safe_year;

        # No need to adjust weekday here, because timelocal()
        # doesn't need weekday in order to compute the number of
        # epoch seconds.
    }

    # 2) Invoke classic timelocal
    my $timelocal = Time::Local::timelocal(@adjusted_time);

    # 3) Add enough seconds to get back
    if( $adjusting ) {
        $timelocal += seconds_between($orig_year, $safe_year);
    }

    return $timelocal;
}


my $days_in_cycle = (365 * 400) + 100 - 4 + 1;
sub seconds_between {
    my($orig, $safe) = @_;

    my $increment = ( $orig > $safe ) ? 1 : -1;
    my $seconds = 0;

    if( $orig > 2400 ) {
        my $cycles = int(($orig - 2400) / 400);
        $orig -= $cycles * 400;
        $seconds += $cycles * $days_in_cycle * 60 * 60 * 24;
    }

    until( $safe == $orig ) {
        my $days = is_leap($safe) ? 366 : 365;
        $seconds += $days * 60 * 60 * 24;
        $safe += $increment;
    }

    return $seconds * $increment + 0; # We prefer 0 to -0.
}


sub is_leap {
    my $year = shift;
    return 1 if $year % 400 == 0;
    return 0 if $year % 100 == 0;
    return 1 if $year % 4   == 0;
}


my @Day_Names   = ("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat");
my @Month_Names = ("Jan", "Feb", "Mar", "Apr", "May", "Jun",
                   "Jul", "Aug", "Sep", "Oct", "Nov", "Dec");

for my $func (qw(localtime gmtime)) {
    my $func64 = do {
        no strict 'refs';
        \&{$func . "64"};
    };

    my $code = sub {
        my $time = @_ ? shift : time;

        my @date = $func64->($time);
        return @date if wantarray;
        return sprintf "%s %s %2d %02d:%02d:%02d %.0f",
          $Day_Names[$date[6]], $Month_Names[$date[4]], $date[3],
          $date[2], $date[1], $date[0],
          $date[5] + 1900;
    };
    
    no strict 'refs';
    *{$func} = $code;
}

sub timegm
{
    my @date = @_[0..5];
    $date[5] = munge_year($date[5]);

    return timegm64(@date);
}

sub UNIX_TIMESTAMP
{
    my $date_time = shift;
    my $unix_timestamp;

    my $year;
    my $month;
    my $day;
    my $hour;
    my $min;
    my $sec;
    if ($date_time =~ /^0000-?00-?00( ?00:?00:?00)?$/)
    {
            return '';
    }
    elsif ($date_time =~ /^(-?\d{4,})-(\d{2})-(\d{2})(?: (\d{2}):(\d{2}):(\d{2}))?$/)
    {
        # DATE or DATETIME
        # "YYYY-MM-DD" or "YYYY-MM-DD hh:mm:ss"
        $year  = $1;
        $month = $2;
        $day   = $3;
        $hour  = $4 || '00';
        $min   = $5 || '00';
        $sec   = $6 || '00';
    }
    elsif ($date_time =~ /^(-?\d{4,}|-?\d{2})(\d{2})(\d{2})(?:(\d{2})(\d{2})(\d{2}))?$/)
    {
        # DATE or DATETIME
        # "YYYYMMDD" or "YYMMDD" or "YYYYMMDDhhmmss" or "YYMMDDhhmmss"
        $year  = $1;
        $month = $2;
        $day   = $3;
        $hour  = $4 || '00';
        $min   = $5 || '00';
        $sec   = $6 || '00';

        if ($year =~ /^\d{2}$/)
        {
            if ($year >= 0 and $year < 38)
            {
                    $year += 2000;
            }
            else
            {
                    $year += 1900;
            }
            warn "Year $year is likely to break something" if $year < 1970;
        }
    }

    my $m = $month - 1;
    my @localtime = ($sec, $min, $hour, $day, $m, $year);

    $unix_timestamp = &timelocal(@localtime);

    return $unix_timestamp;
}

sub FROM_UNIXTIME
{
    my $unix_timestamp = shift;

    if ($unix_timestamp eq '') ### want to warn if undef
    {
        return '0000-00-00 00:00:00';
    }
    elsif ($unix_timestamp !~ /^-?\d+$/)
    {
        cluck "Invalid DATE_TIME '$unix_timestamp'";
        return '0000-00-00 00:00:00';
    }

    my @localtime = &localtime($unix_timestamp);

    my $year  = $localtime[5] + 1900;
    my $month = $localtime[4] + 1;
    my $day   = $localtime[3];
    my $hour  = $localtime[2];
    my $min   = $localtime[1];
    my $sec   = $localtime[0];

    my $date_time = sprintf "%04d-%02d-%02d %02d:%02d:%02d", $year, $month, $day, $hour, $min, $sec;

    return $date_time;
}

1; # of rings to rule them all.

__END__

=head1 NAME

Time::Local::Extended - Increase the range of localtime and timelocal

=head1 SYNOPSIS

  use Time::Local::Extended qw(:ALL);

  my @localtime   = localtime(2**31);
  my $seconds     = timelocal(0,0,0,1,10,170);
  my $gmt_seconds = timegm(0,0,0,1,10,170);
  my $gmt_time    = gmtime(2**31);
  my $ux_time     = UNIX_TIMESTAMP('2097-07-04 12:34:56');
  my $date        = FROM_UNIXTIME(2**31);

  my $sql = qq(
         SELECT start_time
         FROM   project
         WHERE  id = 1
  );
  my $date_time = $dbh->selectrow_array($sql); # '2097-07-04 12:34:56'
  my $ux_time   = UNIX_TIMESTAMP($date_time);  # 4023794096

  my $date_time = FROM_UNIXTIME(2**31);
  my $sql  = qq(
         UPDATE project
         SET    start_time = ?
         WHERE  id = 1
  );
  $dbh->do($sql, undef, $date_time);

=head1 DESCRIPTION

This module extends the date range of localtime(), gmtime(), timegm()
and timelocal() to go safely beyond 2038 and before 1970 on any
operating system.

It also provides a handful of useful time conversion functions.


=head1 PUBLIC FUNCTIONS

=over 4

=item * B<timelocal>

Invoked in the same way as Time::Local::timelocal().

=item * B<localtime>

Invoked in the same way as CORE::localtime().

=item * B<timegm>

Invoked in the same way as Time::Local::timegm().

=item * B<gmtime>

Invoked in the same way as CORE::gmtime().

=item * B<UNIX_TIMESTAMP>

Invoked similarly to the MySQL UNIX_TIMESTAMP() function()

=item * B<FROM_UNIXTIME>

Invoked similarly to the MySQL FROM_UNIXTIME() function()

=back

=head1 LIMITATIONS

Because of the way timegm() and timelocal() try to Do What You Mean
with the year, it is impossible to feed it the years 0 through 99 (it
thinks you mean 2000 - 2099).

While the code can in theory go out to 2**63, the practical portable
limit of this code is from 2**52 to -2**52 (the limit of double
floating point precision) after which precision starts to drop off.
This gives you a range of about +/- 142 million years.

=head1 BUGS

Doesn't correctly handle British Summer Time 1968-1971

Please e-mail bug reports or suggestions to bobo@cpan.org.  Thanks!

=head1 CREDITS

Thanks to Michael Schwern for extending this module's capabilities
beyond 2098.

Thanks to Peter Kioko for helping to refine the idea, and Adam Foxson,
the Human CPAN Reference Manual, for quality assurance.

=head1 AUTHOR

Bob O'Neill, E<lt>bobo@cpan.orgE<gt>
 
=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003-2008 Bob O'Neill and Michael Schwern.
All rights reserved.

See COPYING for license

=head1 SEE ALSO

=over 4

=item * L<perl>.

=item * L<Time::Local>.

=back

=cut
